### ISSUE TITLE
Proposal: Add "filters"

### ISSUE THREAD

**Issue by davidh-ssec** [pytroll/satpy#29]
Currently SatPy has the idea of readers, writers, compositors, and modifiers. I would like to add a "filter" object type. @mraspaud and I have been talking about this on and off for a while so I thought I'd make the discussion a little more official.

A Filter is an optional object or function that when enabled checks the currently loaded datasets against some condition. If the condition is satisfied then the Dataset does not continue on in processing. Otherwise the Dataset continues on as normal. A Filter should *not* modify the values inside a Dataset (only modifiers should do this). However, a Filter may filter out individual pixels if it makes sense.

During our last discussion is was decided that Filters are executed before composites are generated and after composites are generated. I'm not sure whether or not the "before" execution is done inside the readers.

Use cases (please add more as they come up):

1. Day or night filter: useful for reflectances that shouldn't be generated if a scene is more than X% night time. Or 'fog' like composites that are only useful if it is X% night time. How does this differ from per-pixel filtering?
2. Time of year filter: Some products or composites are most useful at a certain time of year. I've seen one case like this with Polar2Grid where there is a certain stretching/enhancement applied to a product that makes it "summer land surface temperature", but another stretching is just "land surface temperature". Not sure if there are others.

**Comment by mraspaud** [comment]
Other use case:
3. Quality filtering: Some data arrive degraded after subnominal reception, The filter would check some parameters (high frequencies in the data for example) and prevent bad quality products from being generated.

**Comment by pnuu** [comment]
I'm not sure this will fit in in the way satpy is implemented (composites are made at load, not for each target area), but I'll add it to the list if you can see a way how to implement it anyway :-)

5. Area filtering: Require at least X % coverage for the target area. Eg. using data from polar satellites not received locally, it is necessary to check how well the data covers the target area so that empty images are not generated. This check/filter should be in very early stage, even before loading data. Can be implemented using _trollsched.satpass.Pass().area_coverage(area_def)_ like in _trollduction.producer.covers()_ and _trollflow_sat.coverage_.

**Comment by mraspaud** [comment]
@pnuu yes, this is a perfectly valid usage for filters.

**Comment by davidh-ssec** [comment]
@mraspaud Isn't this how the filtering works for the `area` parameter to the Scene? It would just need a keyword for percent covered.

**Comment by davidh-ssec** [comment]
6. (maybe 5) Cloud/Sea/Land Masking/Clearing. Using an existing mask in the data files, mask out any data pixels not in this mask or only in this mask.

**Comment by gerritholl** [comment]
Not sure if it's of any use, but fogpy is built around a series of filters: https://github.com/pytroll/fogpy/blob/master/fogpy/filters.py

**Comment by djhoese** [comment]
@gerritholl I think the difference here is that your filters in fogpy are modifying/masking individual pixels. These filters discussed here are meant for per-product filtering ("this product should be produced"). Finding some other name than "filter" this could be a good idea though.

I could see fogpy's filters being considered a "modifier" in Satpy.

**Comment by pnuu** [comment]
Looking at this now, it seems that we now have compositors or enhancements that do something similar to the listed:
1. `DayNightCompositor` (not strictly the same)
3. `MaskingCompositor`
6. `MaskingCompositor`

**Comment by djhoese** [comment]
MaskingCompositor goes inline with @gerritholl's fogpy filters I think, but not with the initial "filtering out datasets from processing" idea of this issue.

**Comment by djhoese** [comment]
FYI I've added the functionality for something like this directly in Polar2Grid/Geo2Grid. I have a sunlight coverage filter (https://github.com/ssec/polar2grid/blob/0d98e9a8286c9e48cd43dc706c40d08c5c1382dd/polar2grid/filters/day_night.py#L101) that is used by two subclasses to say "there is X% of day in this Scene, produce the product" or "X% of night" and whether or 

and a resampling coverage filter (https://github.com/ssec/polar2grid/blob/master/polar2grid/filters/resample_coverage.py#L55) to check the percentage of the output grid that will be covered by the input data, if below a threshold then don't produce it.

**Comment by mraspaud** [comment]
We also have something like this in trollflow2...

**Comment by djhoese** [comment]
Yes, I think as far as this being a "satpy issue", we'll need to decide and define a line for what is useful functionality to Satpy users and how much trollflow2 depends on these other tools for this functionality and how much is trollflow2 specific. I acknowledge that some of these types of feature requests may be out of the scope of Satpy as a **library** but may still be useful. I think we had a similar need for the spherical geometry stuff that was moved from trollschedule to pyresample.
